feat(logger): pretty-print JSON when POWERTOOLS_DEV is set

pmarko1711 opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number: #1527**

## Summary

### Changes

* Logger's default serialize function `json.dumps` will use indentation by four spaces when in AWS SAM Local. Detection happens using [`AWS_SAM_LOCAL` env var](https://github.com/aws/aws-sam-cli/commit/5e56d94c22d71b1ae2a18057ba9182372cade11b) (if `=="true"`).

### User experience

There should be no impact unless in AWS SAM Local. 

In AWS SAM Local, logs are output to console. To make the serialized json messages more readable and consequently debugging with AWS SAM Local easier, logger's default formatter will in this case automatically use indentation by four spaces. That means that messages normally serialized as
```
{"level": "INFO", "location": "lambda_handler:39", "message": "Some log message...", "timestamp": "2022-09-26 09:03:38,349+0000", "service": "issue-1527"}
```
will be in AWS SAM Local by default indented by four spaces:
```
{
    "level": "INFO",
    "location": "lambda_handler:39",
    "message": "Some log message...",
    "timestamp": "2022-09-26 09:02:05,262+0000",
    "service": "issue-1527"
}
This is to make debugging easier
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
    * (only using pytest)
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.


-----
[View rendered docs/core/logger.md](https://github.com/pmarko1711/aws-lambda-powertools-python/blob/develop/docs/core/logger.md)

pmarko1711 committed:
feat(logger): Conditional JSON indentation for AWS SAM Local

heitorlessa commented on aws_lambda_powertools/logging/formatter.py:
```diff
@@ -111,9 +111,15 @@ def __init__(
         self.json_default = json_default or str
-        self.json_serializer = json_serializer or partial(json.dumps, default=self.json_default, separators=(",", ":"))
+        self.json_indent = (
+            4 if os.getenv("AWS_SAM_LOCAL", "").lower() == "true" else None
```
Suggestion: Use a constant to ease maintaining this later, e.g.

```suggestion
            PRETTY_INDENT if os.getenv("AWS_SAM_LOCAL", "").lower() == "true" else COMPACT_INDENT
```

heitorlessa reviewed (changes requested):
Thank you so much for kicking the tires here. Made one suggestion, and I have one important note: let's use `POWERTOOLS_DEV` instead of `AWS_SAM_LOCAL`, I'll create a RFC today.

During maintainers sync, the TypeScript team also had [customers for the same request ](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/751) except that they want to use `node <app.ts>`, or Amplify-CLI, which makes this scenario muddy.

We thought of a new environment variable `POWERTOOLS_DEV` to detect intent on overriding decisions like this and others in the future (e.g., Event Handler has a debug mode which we can deprecate in favour of this in v2). This helps provide a consistent experience across Powertools languages, document behaviours this would enable upfront in our docs, and avoid a proliferation of multiple env vars.

This also futureproof a potential move to AWS Powertools where we could support[1] both Lambda, Fargate, etc.

If your're busy, I'm happy to contribute these changes directly to your fork.

[1] Despite our tenets and name, we've been told by customers who use Powertools in Glue jobs, Fargate, and even IoT; classic [Hyrum's law](https://www.hyrumslaw.com/).

pmarko1711 commented:
OK, I will change this later today. Is there a good place to put the constants in, pls?

heitorlessa commented:
> OK, I will change this later today. Is there a good place to put the constants in, pls?

Yes! `shared/constants.py` is where we keep them.

heitorlessa commented:
We're making progress in the RFC but this might be of interest if others agree this is a clearer intent:

https://github.com/awslabs/aws-lambda-powertools/issues/86#issuecomment-1261126294

pmarko1711 committed:
feat(logger): Conditional JSON indentation on POWERTOOLS_DEV

pmarko1711 commented:
OK, I changed everything to `POWERTOOLS_DEV` and also renamed the test functions and this PR, @heitorlessa .

Since this env var will have multiple effects, shouldn't it be explained in a more prominent place in the documentation? For now I only put an info note in the page about logger. 

Maybe it's not for this PR though.

heitorlessa commented:
thanks a lot @pmarko1711 and yes, I'll create a separate PR [documenting the effects](https://github.com/awslabs/aws-lambda-powertools/issues/86#issuecomment-1262485237).

I'll look into the PR as soon as I complete another task today, hopefully by EOD the latest. With your permission, I might push changes directly to your fork to speed up merging this change (in case I have delays with my primary task) - I'll share feedback nevertheless.

pmarko1711 commented:
ok, yes, no problem @heitorlessa . If you need me to set something (permissions wise) on my fork, let me know.

heitorlessa commented:
Looking at it now ;) Appreciate the patience very much

heitorlessa commented on aws_lambda_powertools/logging/formatter.py:
```diff
@@ -111,9 +111,15 @@ def __init__(
         self.json_default = json_default or str
-        self.json_serializer = json_serializer or partial(json.dumps, default=self.json_default, separators=(",", ":"))
+        self.json_indent = (
+            constants.PRETTY_INDENT if os.getenv("POWERTOOLS_DEV", "").lower() == "true" else constants.COMPACT_INDENT
```
Tip: use [`strtobool`](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/shared/functions.py#L9) as that accounts for variations like `"1"`, `"0"`, etc..

```suggestion
            constants.PRETTY_INDENT if strtobool(os.getenv("POWERTOOLS_DEV", "0")) else constants.COMPACT_INDENT
```

You'll need to import it like

```
from aws_lambda_powertools.shared.functions import strtobool
```

heitorlessa commented on aws_lambda_powertools/shared/constants.py:
```diff
@@ -32,3 +32,6 @@
 ]
+
+PRETTY_INDENT: int = 4
+COMPACT_INDENT = None
```
```suggestion

# JSON indentation level
PRETTY_INDENT: int = 4
COMPACT_INDENT = None
```

Added a quick comment so we don't forget in the future.

heitorlessa commented on docs/core/logger.md:
```diff
@@ -369,6 +369,9 @@ If you prefer configuring it separately, or you'd want to bring this JSON Format
 | **`kwargs`**                 | key-value to be included in log messages                                                                                 | `None`                                                        |
 
+???+ info
+    When `POWERTOOLS_DEV` env var is present and set to `"true"`, Logger's default serializer (`json.dumps`) will pretty-print log messages for easier readability.
```
Great wording!

pmarko1711 replied:
I am not a native English speaker, so many thanks

heitorlessa commented on tests/functional/test_logger_powertools_formatter.py:
```diff
@@ -288,3 +290,22 @@ def test_log_formatting(stdout, service_name):
+def test_log_json_indent_compact_indent(stdout, service_name, monkeypatch):
+    # GIVEN a logger with default settings and WHEN POWERTOOLS_DEV is not set
+    if "POWERTOOLS_DEV" in os.environ:
+        monkeypatch.delenv(name="POWERTOOLS_DEV")
```
```suggestion
    monkeypatch.delenv(name="POWERTOOLS_DEV", raising=False)
```

[`delenv`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.MonkeyPatch.delenv) supports a flag to not raise an exception if the env var doesn't exist.

heitorlessa commented on tests/functional/test_logger_powertools_formatter.py:
```diff
@@ -288,3 +290,22 @@ def test_log_formatting(stdout, service_name):
+    logger = Logger(service=service_name, stream=stdout)
+    logger.info("Test message")
+    # THEN the json should not be indented using constant.PRETTY_INDENT blank spaces
+    assert " " * constants.PRETTY_INDENT not in stdout.getvalue()
```
```suggestion
    new_lines = stdout.getvalue().count("\n")
    assert new_lines == 1
```

Tip: non-indented log lines should have a single `\n`. While this assertion works for this test, it isn't always correct in the `test_log_json_pretty_indent` test because `assert " " * 0 in stdout.getvalue()` is also True.

pmarko1711 replied:
Hmm, at first I thought the same but then this test failed. For some reason,there was one new line (but I think my message at the time was two words, "Test message" and for some reason it got split in the middle.)

pmarko1711 replied:
also, should we not rather use `os.linesep`  instead of `\n`?

pmarko1711 replied:
I've commited your suggestion, but using `os.linesep`

heitorlessa commented on tests/functional/test_logger_powertools_formatter.py:
```diff
@@ -288,3 +290,22 @@ def test_log_formatting(stdout, service_name):
+    logger = Logger(service=service_name, stream=stdout)
+    logger.info("Test message")
+    # THEN the json should contain indentation (of constant.PRETTY_INDENT blank spaces)
+    assert " " * constants.PRETTY_INDENT in stdout.getvalue()
```
```suggestion
    new_lines = stdout.getvalue().count("\n")
    assert new_lines > 1
```

See my earlier comment. You could also do `assert new_lines == 7` if you want to be 100% precise, but it's brittle if we add additional keys.

pmarko1711 replied:
I've commited your suggestion, but using os.linesep

heitorlessa reviewed (changes requested):
Nice work!! I love the wording - straight to the point.

Made a few suggestions to improve test accuracy, and use our `strtobool` to support cases where customers may use boolean strings like `0` or `1`

pmarko1711 committed:
Update aws_lambda_powertools/logging/formatter.py

Co-authored-by: Heitor Lessa <lessa@amazon.nl>

pmarko1711 committed:
Update aws_lambda_powertools/shared/constants.py

Co-authored-by: Heitor Lessa <lessa@amazon.nl>

pmarko1711 committed:
Update tests/functional/test_logger_powertools_formatter.py

Co-authored-by: Heitor Lessa <lessa@amazon.nl>

pmarko1711 committed:
Tests counting newlines only

pmarko1711 commented:
I've committed all your suggestions, @heitorlessa , with a small change - I counted `os.linesep` instead of `\n` in the functional tests

heitorlessa commented:
Looking into now ;) Thank you so much for these changes!

heitorlessa commented:
Looks great @pmarko1711 - thank you so much for working throughout the review process. Merging now.

**Next steps**: I'll **(1)** update the RFC with the latest info on `POWERTOOLS_DEV` and `POWERTOOLS_DEBUG` (separate feat altogether), **(2)** create a PR to update the documentation on these flags, and **(3)** release this feature by the end of the week.

pmarko1711 commented:
thanks for all the support, @heitorlessa
